src/main.rs

```diff
@@ -46,12 +47,72 @@ use libtock_drivers::console::Console;
 use libtock_drivers::result::FlexUnwrap;
 use libtock_drivers::timer::Duration;
 use libtock_drivers::usb_ctap_hid;
+use usb_ctap_hid::UsbEndpoint;
 
 libtock_core::stack_size! {0x4000}
 
 const SEND_TIMEOUT: Milliseconds<ClockInt> = Milliseconds(1000);
 const KEEPALIVE_DELAY_TOCK: Duration<isize> = Duration::from_ms(KEEPALIVE_DELAY_MS as isize);
 
+#[cfg(not(feature = "vendor_hid"))]
+const NUM_ENDPOINTS: usize = 1;
+#[cfg(feature = "vendor_hid")]
+const NUM_ENDPOINTS: usize = 2;
+
+// The reply/replies that are queued for each endpoint.
+struct EndpointReply {
+    endpoint: UsbEndpoint,
+    transport: Transport,
+    reply: HidPacketIterator,
+}
+
+impl EndpointReply {
+    pub fn new(endpoint: UsbEndpoint) -> Self {
+        EndpointReply {
+            endpoint,
+            transport: match endpoint {
+                UsbEndpoint::MainHid => Transport::MainHid,
+                #[cfg(feature = "vendor_hid")]
+                UsbEndpoint::VendorHid => Transport::VendorHid,
+            },
+            reply: HidPacketIterator::none(),
+        }
+    }
+}
+
+// A single packet to send.
+struct SendPacket {
+    transport: Transport,
+    packet: [u8; 64],
+}
+
+struct Endpoints {
+    endpoints: [EndpointReply; NUM_ENDPOINTS],
+}
+
+impl Endpoints {
+    pub fn new() -> Self {
+        Endpoints {
+            endpoints: [
+                EndpointReply::new(UsbEndpoint::MainHid),
+                #[cfg(feature = "vendor_hid")]
+                EndpointReply::new(UsbEndpoint::VendorHid),
+            ],
+        }
+    }
+
+    pub fn next_packet(&mut self) -> Option<SendPacket> {
+        for ep in self.endpoints.iter_mut() {
```

Comment:
I guess round-robin is not necessary here?

Reply:
Not necessary to get the tests to pass. But this doesn't test on host behavior.  It could be done. 
Round robin will:
+ ensure latency between packets for each endpoint is kept small for the host
+ a small reply won't get blocked by a large reply as much
- increase time to send entire reply for the first endpoint

Current approach of Main HID first:
+ prioritises Main HID replies (arguably more important for users)
+ simpler
- delays Vendor HID, even if it small it will be sent after a large Main HID replay.
- may introduce large delays between replies for Vendor HID.

What do you think?

Reply:
I think Vendor HID is more important because you don't want updates to get interrupted. But the correct solution is probably to disable the Main HID during an update.

I'm fine to keep as is if there's no current issue and deal with it later when designing upgradability or when an issue arises.